# Lending Club Case Study

# Problem Statement
Working for a consumer finance company which specialises in lending various types of loans to urban customers. When the company receives a loan application, the company has to make a decision for loan approval based on the applicant’s profile. Two types of risks are associated with the bank’s decision:
<br>
    1) If the applicant is likely to repay the loan, then not approving the loan results in a loss of business to the company
<br>
    2) If the applicant is not likely to repay the loan, i.e. he/she is likely to default, then approving the loan may lead to a financial loss for the company
    <br>
<br>
The data given to us contains information about past loan applicants and whether they ‘defaulted’ or not. 
<br>
The aim is to identify patterns which indicate if a person is likely to default, which may be used for taking actions such as denying the loan, reducing the amount of loan, lending (to risky applicants) at a higher interest rate, etc.
 <br>
 <br>
If one is able to identify these risky loan applicants, then such loans can be reduced thereby cutting down the amount of credit loss. Identification of such applicants using EDA is the aim of this case study
<br>
 <br>
Company wants to understand the driving factors (or driver variables) behind loan default, i.e. the variables which are strong indicators of default.  The company can utilise this knowledge for its portfolio and risk assessment. 



Importing the libraries

In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings 
warnings.filterwarnings('ignore') 

In [24]:
lend_data=pd.read_csv("loan.csv")
lend_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 111 columns]

In [85]:
#Find the number of rows and columns
lend_data.shape

(39717, 111)

We hav 111 columns and 39717 rows.
<br>

Find the number of attributes which has all null values and drop the same.

In [86]:
#Find columns which has all null values and count the same
c=0
for i in lend_data.columns:
    if lend_data[i].isnull().sum()==39717:
        print(i)
        c+=1
print("Total number of columns which has all null values: ",c)

mths_since_last_major_derog
annual_inc_joint
dti_joint
verification_status_joint
tot_coll_amt
tot_cur_bal
open_acc_6m
open_il_6m
open_il_12m
open_il_24m
mths_since_rcnt_il
total_bal_il
il_util
open_rv_12m
open_rv_24m
max_bal_bc
all_util
total_rev_hi_lim
inq_fi
total_cu_tl
inq_last_12m
acc_open_past_24mths
avg_cur_bal
bc_open_to_buy
bc_util
mo_sin_old_il_acct
mo_sin_old_rev_tl_op
mo_sin_rcnt_rev_tl_op
mo_sin_rcnt_tl
mort_acc
mths_since_recent_bc
mths_since_recent_bc_dlq
mths_since_recent_inq
mths_since_recent_revol_delinq
num_accts_ever_120_pd
num_actv_bc_tl
num_actv_rev_tl
num_bc_sats
num_bc_tl
num_il_tl
num_op_rev_tl
num_rev_accts
num_rev_tl_bal_gt_0
num_sats
num_tl_120dpd_2m
num_tl_30dpd
num_tl_90g_dpd_24m
num_tl_op_past_12m
pct_tl_nvr_dlq
percent_bc_gt_75
tot_hi_cred_lim
total_bal_ex_mort
total_bc_limit
total_il_high_credit_limit
Total number of columns which has all null values:  54


We can drop all 54 columns as these will not add values for our analysis


In [87]:
lend_data.dropna(axis=1, how="all", inplace=True)  
# axis = 0 -> row, axis = 1 -> column

In [88]:
lend_data.shape

(39717, 57)

Now we have 57 columns for the analysis.


Find the columns which has only one value

In [89]:
c=0
for i in lend_data.columns:
    if lend_data[i].nunique()==1:
        print(i)
        c+=1
print(c)

pymnt_plan
initial_list_status
collections_12_mths_ex_med
policy_code
application_type
acc_now_delinq
chargeoff_within_12_mths
delinq_amnt
tax_liens
9


Above 9 variables are having single value or in combination with NA.<br>
Variables like collections_12_mths_ex_med, chargeoff_within_12_mths,tax_liens etc has value 0 with NA. As both values are not useful for our anlysis, we will drop these

In [90]:
lend_data.drop(['pymnt_plan', "initial_list_status",'collections_12_mths_ex_med','policy_code','application_type', 'acc_now_delinq', 'chargeoff_within_12_mths', 'delinq_amnt', 'tax_liens'], axis = 1, inplace = True)
lend_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... total_rec_prncp total_rec_int  \
0   10.65%       162.87     B        B2  ...         5000.00        863.16   
1   15.27%        59.83     C        C4  ...          456.46        435.17   
2   15.96%        84.33     C        C5  ...         2400.00        605.67   
3   13.49%       339.31     C        C1  ...        10000.00       2214.92   
4   12.69%        67.79     B        B5  ...         2475.94       1037.39   

  total_rec_late_fee  recoveries collection_recovery_fee last_pymnt_d  \
0               0.00        0.00                    0.00       Jan-15   
1               0.00      117.08                    1.11       Apr-13   
2               0.00        0.00                    0.00       Jun-14   
3              16.97        0.00                    0.00       Jan-15   
4               0.00        0.00                    0.00       May-16   

  last_pymnt_amnt next_pymnt_d last_credit_pull_d pub_rec_bankruptcies  
0          171.62          NaN             May-16                  0.0  
1          119.66          NaN             Sep-13                  0.0  
2          649.91          NaN             May-16                  0.0  
3          357.48          NaN             Apr-16                  0.0  
4           67.79       Jun-16             May-16                  0.0  

[5 rows x 48 columns]

In [91]:
lend_data.shape

(39717, 48)

Now we have 48 columns. Lets understand the features of each columns

In [92]:
lend_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 48 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39717 non-null  int64  
 1   member_id                39717 non-null  int64  
 2   loan_amnt                39717 non-null  int64  
 3   funded_amnt              39717 non-null  int64  
 4   funded_amnt_inv          39717 non-null  float64
 5   term                     39717 non-null  object 
 6   int_rate                 39717 non-null  object 
 7   installment              39717 non-null  float64
 8   grade                    39717 non-null  object 
 9   sub_grade                39717 non-null  object 
 10  emp_title                37258 non-null  object 
 11  emp_length               38642 non-null  object 
 12  home_ownership           39717 non-null  object 
 13  annual_inc               39717 non-null  float64
 14  verification_status   

We will analyze the object columns first. <br> 
Finding value count for all object variables

In [93]:
for i in lend_data.columns[lend_data.dtypes == 'object']:
    print(lend_data[i].value_counts())
    print('-----------------------------------------------')

 36 months    29096
 60 months    10621
Name: term, dtype: int64
-----------------------------------------------
10.99%    956
13.49%    826
11.49%    825
7.51%     787
7.88%     725
         ... 
18.36%      1
16.96%      1
16.15%      1
16.01%      1
17.44%      1
Name: int_rate, Length: 371, dtype: int64
-----------------------------------------------
B    12020
A    10085
C     8098
D     5307
E     2842
F     1049
G      316
Name: grade, dtype: int64
-----------------------------------------------
B3    2917
A4    2886
A5    2742
B5    2704
B4    2512
C1    2136
B2    2057
C2    2011
B1    1830
A3    1810
C3    1529
A2    1508
D2    1348
C4    1236
C5    1186
D3    1173
A1    1139
D4     981
D1     931
D5     874
E1     763
E2     656
E3     553
E4     454
E5     416
F1     329
F2     249
F3     185
F4     168
F5     118
G1     104
G2      78
G4      56
G3      48
G5      30
Name: sub_grade, dtype: int64
-----------------------------------------------
US Army                      

 Term , int_rate,revol_util,emp_length has some strings and need to convert to numeric value

In [94]:
lend_data["term"]=lend_data["term"].str.replace(" months","").astype("int")

In [95]:
lend_data["int_rate"]=lend_data["int_rate"].str.replace("%","").astype("float")

In [96]:
lend_data["revol_util"]=lend_data["revol_util"].str.replace("%","").astype("float")

In [123]:
lend_data['emp_length'].isnull().sum()/lend_data['emp_length'].count()

0.02781947104187154

Only ~2.7% of data has null values and hence replace null value with mode value

In [124]:
lend_data['emp_length'] = lend_data['emp_length'].fillna(lend_data['emp_length'].mode()[0])

In [128]:
#emp_length column convert to int considering 0 means less than one year and 10 means ten or more years
lend_data['emp_length'] = lend_data['emp_length'].map(lambda x:x.split('+ years')[0] if x == '10+ years' else ('0' if x == '< 1 year' else(x.strip(' years')))).astype(int)
